### PR TITLE
add auth, logout, jwt

### PR DESCRIPTION
**Pull request by darkhanzhursin**
https://trello.com/c/M4oFoyRC/28-be-users-finish-up-setting-routes

### CODE REVIEW

**Review by DonJuvn — APPROVED** [review]

File: server/routes/auth.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+const router = require('express').Router();
+const jwt = require('jsonwebtoken');
+const Users = require('../models/Users');
+const bcrypt = require('bcrypt');
+
+router.post('/', async (req, res) => {
+  const { email, pwd } = req.body;
+  if (!email || !pwd) {
```
Comment (Daulet-29) [inline server/routes/auth.js:9]: думаю так лучше будет if (!email && !pwd) {

**Review by Daulet-29 — APPROVED** [review]

File: server/routes/auth.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
+const router = require('express').Router();
+const jwt = require('jsonwebtoken');
+const Users = require('../models/Users');
```
Comment (eyesofdaveed) [inline server/routes/auth.js:4]: import orders

Comment (darkhanzhursin) [reply server/routes/auth.js:4]: fixed

File: server/routes/auth.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    res.cookie('jwt', refreshToken, {
+      httpOnly: true,
+      sameSite: 'None',
+      // secure: true,
+      maxAge: 24 * 60 * 60 * 1000,
+    });
+    res.json({
+      accessToken,
+      message: `Successfully authorized: ${foundUser.email} ${foundUser.firstName} ${foundUser.lastName}`,
```
Comment (eyesofdaveed) [inline server/routes/auth.js:53]: success: true,

File: server/routes/logout.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+  // Is refreshToken in db?
+  const foundUser = await Users.findOne({ refreshToken }).exec();
+  if (!foundUser) {
+    res.clearCookie('jwt', { httpOnly: true, sameSite: 'None', secure: true });
+    return res.sendStatus(204);
+  }
+
+  // Delete refreshToken in db
+  foundUser.refreshToken = '';
```
Comment (eyesofdaveed) [inline server/routes/logout.js]: const { refreshToken, ...updateUser} = foundUser;

File: server/routes/auth.js (hunk 3 above)
Comment (darkhanzhursin) [reply server/routes/auth.js:53]: fixed

File: server/routes/logout.js (hunk 4 above)
Comment (darkhanzhursin) [reply server/routes/logout.js]: fixed

File: server/routes/users.js [hunk 5]
```diff
@@ -1,45 +1,58 @@
...
-})
+  const { email, password } = req.body;
+  if (!email) return res.status(400).json({ message: 'Email is required.' });
+
+  if (!password)
+    return res.status(400).json({ message: 'Password is required.' });
+
+  // check for duplicate email in the db
+  const duplicateEmail = await Users.findOne({ email: email }).exec();
+  if (duplicateEmail) return res.sendStatus(409); //Conflict
```
Comment (eyesofdaveed) [inline server/routes/users.js]: inside of try block

File: server/routes/users.js [hunk 6]
```diff
@@ -1,45 +1,58 @@
...
+
+  try {
+    //encrypt the password
+    const hashedPwd = await bcrypt.hash(password, hashConstance);
+    const newUser = new Users({
+      email: email,
```
Comment (eyesofdaveed) [inline server/routes/users.js]: email,

File: server/routes/users.js [hunk 7]
```diff
@@ -1,45 +1,58 @@
...
-    }
-})
+  const { email, password } = req.body;
+  if (!email) return res.status(400).json({ message: 'Email is required.' });
+
+  if (!password)
+    return res.status(400).json({ message: 'Password is required.' });
+
+  // check for duplicate email in the db
+  const duplicateEmail = await Users.findOne({ email: email }).exec();
```
Comment (eyesofdaveed) [inline server/routes/users.js]: email,

File: server/routes/users.js [hunk 8]
```diff
@@ -1,45 +1,58 @@
...
-        });
-
-        const user = await newUser.save();
-        res.status(200).json(user);
-    } catch (err) {
-        console.log(err);
-    }
-})
+  const { email, password } = req.body;
+  if (!email) return res.status(400).json({ message: 'Email is required.' });
```
Comment (eyesofdaveed) [inline server/routes/users.js]: succes: false,

File: server/routes/logout.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
+const Users = require('../models/Users');
+
+const router = require('express').Router();
+
+router.get('/', async (req, res) => {
+  const cookies = req.cookies;
+  if (!cookies?.jwt) return res.sendStatus(204); // No content
+  const refreshToken = cookies.jwt;
+
+  const filter = { refreshToken: refreshToken };
```
Comment (eyesofdaveed) [inline server/routes/logout.js]: refreshToken

File: server/routes/logout.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  const updatedUser = await Users.findOneAndUpdate(filter, update, {
+    new: true,
+  });
+  if (!updatedUser) {
+    res.clearCookie('jwt', { httpOnly: true, sameSite: 'None', secure: true });
+    return res.sendStatus(204);
+  }
+
+  const result = await updatedUser.save();
+  console.log(result);
```
Comment (eyesofdaveed) [inline server/routes/logout.js]: remove log

File: server/routes/logout.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
+const Users = require('../models/Users');
+
+const router = require('express').Router();
+
+router.get('/', async (req, res) => {
+  const cookies = req.cookies;
+  if (!cookies?.jwt) return res.sendStatus(204); // No content
```
Comment (eyesofdaveed) [inline server/routes/logout.js]: separate logic

File: server/routes/auth.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    return res
+      .status(400)
+      .json({ message: 'Email and password are required.' });
+  }
+
+  const foundUser = await Users.findOne({ email: email }).exec();
+
+  //Unauthorized
+  if (!foundUser) return res.sendStatus(401);
+
```
Comment (eyesofdaveed) [inline server/routes/auth.js:19]: const { email, firstName } = foundUser;

File: server/routes/auth.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  const match = await bcrypt.compare(pwd, foundUser.password);
+  if (match) {
+    // create JWTs
+    const accessToken = jwt.sign(
+      {
+        UserInfo: {
+          email: foundUser.email,
```
Comment (eyesofdaveed) [inline server/routes/auth.js]: email, firstName

**Comment by Bashkanitto** [comment]
fix conflicts please, there are mistaces cause of this,

Resolve conflicts before merging

File: server/routes/logout.js (hunk 11 above)
Comment (darkhanzhursin) [reply server/routes/logout.js]: fixed

File: server/routes/logout.js (hunk 10 above)
Comment (darkhanzhursin) [reply server/routes/logout.js]: fixed

File: server/routes/users.js (hunk 8 above)
Comment (darkhanzhursin) [reply server/routes/users.js]: fixed

File: server/routes/users.js (hunk 7 above)
Comment (darkhanzhursin) [reply server/routes/users.js]: added

File: server/routes/users.js (hunk 5 above)
Comment (darkhanzhursin) [reply server/routes/users.js]: added

File: server/routes/users.js (hunk 6 above)
Comment (darkhanzhursin) [reply server/routes/users.js]: fixed

File: server/routes/auth.js (hunk 13 above)
Comment (darkhanzhursin) [reply server/routes/auth.js]: fixed

**Review by eyesofdaveed — APPROVED** [review]

**Pull request merged**